plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/GroovySourceFileAllowlist.java

```diff
@@ -152,6 +148,22 @@ private static boolean endsWithIgnoreCase(String value, String suffix) {
         }
     }
 
+    @Extension
+    public static class AllowedGroovyResourcesCache {
+        private final Map<String, Boolean> cache = new ConcurrentHashMap<>();
+
+        public boolean isAllowed(String groovySourceFileUrl) {
+            return cache.computeIfAbsent(groovySourceFileUrl, url -> {
+                for (GroovySourceFileAllowlist allowlist : GroovySourceFileAllowlist.all()) {
+                    if (allowlist.isAllowed(url)) {
+                        return true;
+                    }
+                }
+                return false;
```

Comment:
I do not think we want to cache negative results? Unclear to me if this would cause problems with dynamic plugin installation with https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/686.

Reply:
Admittedly only if some weird pipeline attempted to load a plugin-supplied source file directly, and _before_ that plugin was loaded. Anyway, we do not expect allowlists to reject anything under normal operation so it should not be necessary to optimize for that case right?

Reply:
> Admittedly only if some weird pipeline attempted to load a plugin-supplied source file directly, and before that plugin was loaded.

Yeah, I played around with this case when I filed the PR, and I think you would have to be doing something quite unusual in your Pipeline to run into problems due to caching negative results. 

> Anyway, we do not expect allowlists to reject anything under normal operation so it should not be necessary to optimize for that case right?

Sure, fair enough.

Reply:
See [777deff](https://github.com/jenkinsci/workflow-cps-plugin/pull/829/commits/777deff649dbd81b45f63bbdd4d341bb7a473824).